src/core/smp.cc

```diff
@@ -148,4 +150,89 @@ smp_service_group_semaphore& get_smp_service_groups_semaphore(unsigned ssg_id, s
     return smp_service_groups[ssg_id].clients[t];
 }
 
+smp::smp(alien::instance& alien)
+        : _alien(alien) {
+}
+
+
+smp::~smp() = default;
+
+void
+smp::setup_prefaulter(seastar::resource::hwloc::internal::topology_holder& topo, seastar::memory::internal::numa_layout layout) {
+    // Stack guards mprotect() random pages, so the prefaulter will hard-fault.
+#ifndef SEASTAR_THREAD_STACK_GUARDS
+    _prefaulter = std::make_unique<internal::memory_prefaulter>(topo, std::move(layout));
+#endif
+}
+
+internal::memory_prefaulter::memory_prefaulter(resource::hwloc::internal::topology_holder& topo, memory::internal::numa_layout layout) {
+    for (auto& range : layout.ranges) {
+        _layout_by_node_id[range.numa_node_id].push_back(std::move(range));
+    }
+    auto page_size = getpagesize();
+    for (auto&& [numa_node_id, ranges] : _layout_by_node_id) {
+        posix_thread::attr a;
+        auto cpus_for_numa_node = resource::internal::cpus_for_numa_node_id(topo, numa_node_id);
+        if (cpus_for_numa_node) {
+            cpu_set_t cpuset;
+            CPU_ZERO(&cpuset);
+            for (auto cpu : *cpus_for_numa_node) {
+                CPU_SET(cpu, &cpuset);
+            }
+            a.set(cpuset);
+        };
+        _worker_threads.emplace_back(a, [this, numa_node_id = numa_node_id, &ranges, page_size] {
+            work(ranges, page_size);
+        });
+    }
+}
+
+internal::memory_prefaulter::~memory_prefaulter() {
+    _stop_request.test_and_set(std::memory_order_relaxed);
+    for (auto& t : _worker_threads) {
+        t.join();
+    }
+}
+
+void
+internal::memory_prefaulter::work(std::vector<memory::internal::memory_range>& ranges, size_t page_size) {
+    sched_param param = { .sched_priority = 0 };
+    // SCHED_IDLE doesn't work via thread attributes
+    pthread_setschedparam(pthread_self(), SCHED_IDLE, &param);
+    size_t current_range = 0;
+    const size_t batch_size = 512; // happens to match huge page size on x86, through not critical
+    while (!_stop_request.test(std::memory_order_relaxed) && !ranges.empty()) {
+        auto& range = ranges[current_range];
+        // copy eveything into locals, or the optimizer will worry about them due
+        // to the cast below and not optimize anything
+        auto start = range.start;
+        auto end = range.end;
+        for (size_t i = 0; i < batch_size && start < end; ++i) {
+            // atomic_ref would be better, but alas C++20 only
+            auto p = reinterpret_cast<volatile std::atomic<char>*>(start);
+            // Touch the page for write, but be sure not to modify anything
+            p->fetch_or(0, std::memory_order_relaxed);
+            start += page_size;
```

Comment:
What's the point in touching the whole huge-page with page-size steps? If the first (well, zeroth) touch allocates the huge-page, the we're touching the remaining 511 steps in vain. If it doesn't it means that we're touching small pages and delay huge-page split into kernel's background anyway

Reply:
How do we know the compiler wouldn't outsmart us and since we do not actually do anything there, it won't just eliminate this? Something like https://owasp.org/www-community/vulnerabilities/Insecure_Compiler_Optimization ?

Reply:
It's marked as `volatile` while casting, so it shouldn't

Reply:
> What's the point in touching the whole huge-page with page-size steps? If the first (well, zeroth) touch allocates the huge-page, the we're touching the remaining 511 steps in vain. If it doesn't it means that we're touching small pages and delay huge-page split into kernel's background anyway

The thinking was that we don't know if huge pages work or not, it could be disabled in the kernel.

If huge pages work, then 511 extra atomics are meaningless compared to the page faults (~25 cycle for an atomic compared to at least 4000 cycles for the page fault + lots more cycles for the zeroing + L1/L2 caches destroyed).

If they don't work, then the 511 extra atomics prevent scylladb/scylladb#8828.

btw, I saw patches to use non-temporal stores to zero huge pages, but they didn't make it in.